### PR TITLE
Add new LLVM JITLink blog post

### PR DESCRIPTION
**Pull request by jkshtj**
This commit adds a blog post about adding a new target/object backend to
LLVM JITLink.

### CODE REVIEW

**Comment by vgvassilev** [comment]
I reviewed a draft of this and seemed quite exciting. Maybe @lhames and @weliveindetail could give their opinion?

**Comment by jkshtj** [comment]
I'll let @lhames comment separately as well, but I did review this with him offline. Would still love feedback from @weliveindetail 🙂.

File: content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md
Comment (lhames) [inline content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md:100]: I missed this earlier, but for scriptable extensions I think the idea is that you can run a script (under the JIT) against some API surface you expose from your app.

The classic way to extend a precompiled program dynamically is with plugins that are compiled to a shared library and then dlopened. Alternatively people add bindings for some scripting language (e.g. python) and let plugin writers use that, but then you need to manage the foreign function interface between your program's native language / ABI and the scripting language. The nice thing about JIT linking is that you could use it as a basis for writing scripts in the native language of the program.

File: content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md
Comment (lhames) [inline content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md:131]: > "...parse ELF files, from scratch..."
No comma needed here.

**Review by lhames — COMMENTED** [review]
@jkshtj and I have been discussing this offline while he drafted it. I think it's a great write up of his project, and would be a useful resource for people looking to write a JITLink backend, or who are interested in JITLink/ORC in general.

I found some minor typos while re-reading, and one piece of text that seemed out of place, but otherwise LGTM!

File: content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md
Comment (jkshtj) [reply content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md:33]: Fixed.

File: content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md
Comment (jkshtj) [reply content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md:89]: Fixed.

File: content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md
Comment (jkshtj) [reply content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md:131]: Fixed.

File: content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md
Comment (jkshtj) [reply content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md:100]: As per offline discussion changed this to -
>  **Scriptable extensions** (think about running JIT’d code in the context of some existing app, allowing the app to be extended by JIT’d code rather than precompiled plugins)

**Comment by lhames** [comment]
Looks good to me. Thanks @jkshtj!

**Review by lhames — APPROVED** [review]

**Comment by jkshtj** [comment]
Thanks @lhames! @vgvassilev can we merge this 🙂?

**Comment by weliveindetail** [comment]
That's a nice write-up. Thanks for working on it! I read through 1/3rd I believe and left some notes.

**Comment by jkshtj** [comment]
@weliveindetail are you sure you published your comments? I don't see anything from you.

**Review by weliveindetail — COMMENTED** [review]
Sorry, didn't submit my comments. Here we go.

File: content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md
Comment (jkshtj) [reply content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md:107]: I see that it may not be inherently clear what I meant by low-level. What I really meant to say was that it is lower level than ORC, which is, sort of, the public API for JIT linking in LLVM. Perhaps I can be explicit here and say that instead, so that the meaning of "low-level" is clear.

File: content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md
Comment (jkshtj) [reply content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md:31]: Fixed.

File: content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md
Comment (jkshtj) [reply content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md:109]: Fixed.

File: content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md
Comment (jkshtj) [reply content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md:110]: Fixed. Also introduced the term "executor process" earlier so it doesn't seem like we're dropping it out of nowhere.

File: content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md
Comment (jkshtj) [reply content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md:146]: Yep. Fixed.

File: content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md
Comment (jkshtj) [reply content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md:141]: Good catch! A number of github links were already broken.

File: content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md
Comment (jkshtj) [reply content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md:163]: I did mean "and". Fixed it.

File: content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md
Comment (jkshtj) [reply content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md:169]: > You could make space by reducing the higher level concepts into one sentence, since they all map to ELF very well.

I think we would be assuming that the reader or a new contributor is already familiar with ELF or object formats in general. For that reason I'd prefer being a bit verbose here.

**Comment by jkshtj** [comment]
@weliveindetail published a new commit, addressing your feedback. Thank you for taking a look!

**Comment by weliveindetail** [comment]
You are welcome. I cross-read through the rest of the post. I think it's ok, but TBH from an uninvolved reader's perspective it's a bit too much storytelling. I can perfectly understand that it's your project and you want to tell how you managed to get it done, but take care not to lose the reader on the way.

In this paragraph e.g. you say it was tricky and you got it done, but what's the take-away fro the reader? (You had to cross-compile? Why? How? Were there alternative? etc.)
```
The tricky part here, at least for me, was to get a 32-bit i386 llvm-jitlink executable. By default Clang produces
executables for the host architecture, which meant that I needed to dabble a bit in cross-compilation[^5] (compiling
for a target different than the host architecture) since I was developing on x86-64[^6] hardware. But eventually, I
had on my hands a 32-bit i386 llvm-jitlink executable!
```

Rule of thumb might be: Either keep it a note of 1 sentence max OR discuss the entire problem: which options did you have, what you decided for, how you did it (ideally with concrete code/commands). That way you force yourself to keep a focus. Otherwise you risk to describe **everything** on a shallow niveau, but IMHO that's not interesting for most readers.

I made a few more notes that are not exhaustive. Maybe go through yourself and see where you can compress things a bit? Otherwise, as I said, I think you can post it like that. It doesn't need to be perfect as long as it's correct from the technical point of view. Anyway, I didn't want to hold back with this remark.

**Comment by jkshtj** [comment]
> You are welcome. I cross-read through the rest of the post. I think it's ok, but TBH from an uninvolved reader's perspective it's a bit too much storytelling. I can perfectly understand that it's your project and you want to tell how you managed to get it done, but take care not to lose the reader on the way.
> 
> In this paragraph e.g. you say it was tricky and you got it done, but what's the take-away fro the reader? (You had to cross-compile? Why? How? Were there alternative? etc.)
> 
> ```
> The tricky part here, at least for me, was to get a 32-bit i386 llvm-jitlink executable. By default Clang produces
> executables for the host architecture, which meant that I needed to dabble a bit in cross-compilation[^5] (compiling
> for a target different than the host architecture) since I was developing on x86-64[^6] hardware. But eventually, I
> had on my hands a 32-bit i386 llvm-jitlink executable!
> ```
> 
> Rule of thumb might be: Either keep it a note of 1 sentence max OR discuss the entire problem: which options did you have, what you decided for, how you did it (ideally with concrete code/commands). That way you force yourself to keep a focus. Otherwise you risk to describe **everything** on a shallow niveau, but IMHO that's not interesting for most readers.
> 

Thank you for this recommendation Stefan! I went over the post again and trimmed it down, so that it's more objective. I like how it reads now, much better.

File: content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md
Comment (jkshtj) [reply content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md]: Fixed in latest commit.

File: content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md
Comment (lhames) [reply content/posts/2023-03-16-Adding-new-LLVM-JITLink-target-object-backend.md:67]: The code below is C++. :)

**Comment by lattner** [comment]
It seems like this is converging?  If so, please let me know and I'll do a pass over it.  Thanks!

**Comment by jkshtj** [comment]
I believe so. We had an offline conversation about this and at least @lhames was good with landing this. I'll let him and others comment separately. Thank you!

**Comment by lattner** [comment]
The blog post looks nice to me.  @lhames assuming you are fine with it, I'll merge it tomorrow!  Thank you for writing this, great work!

**Review by weliveindetail — APPROVED** [review]
Thanks for your update @jkshtj, I agree that it's in a good shape now!

**Comment by lhames** [comment]
@lattner Yep -- Looks good to me.

@jkshtj Thank you again for your work on this, and on the i386 backend -- they're great contributions!

**Comment by lattner** [comment]
Amazing, it's live!  Congratulations!!

**Comment by jkshtj** [comment]
@lattner looks like the build failed. I see the following error - 
```

This is a scheduled Ubuntu-18.04 brownout. The Ubuntu-18.04 environment is deprecated and will be removed on April 1st, 2023. For more details, see https://github.com/actions/runner-images/issues/6002

```

**Comment by lattner** [comment]
Looks like it worked:
https://blog.llvm.org/posts/2023-03-16-adding-new-llvm-jitlink-target-object-backend/

Congrats again and thank you for the great post!

**Comment by tlattner** [comment]
I fixed the GitHub actions. Also, we need to check the date is right before merging  as another post happened while this was reviewed.-TanyaOn Mar 28, 2023, at 9:21 PM, Chris Lattner ***@***.***> wrote:﻿
Looks like it worked:
https://blog.llvm.org/posts/2023-03-16-adding-new-llvm-jitlink-target-object-backend/
Congrats again and thank you for the great post!

**Comment by tlattner** [comment]
Also, thanks for the amazing blog post!-TanyaOn Mar 28, 2023, at 9:44 PM, Tanya Lattner ***@***.***> wrote:﻿I fixed the GitHub actions. Also, we need to check the date is right before merging  as another post happened while this was reviewed.-TanyaOn Mar 28, 2023, at 9:21 PM, Chris Lattner ***@***.***> wrote:﻿
Looks like it worked:
https://blog.llvm.org/posts/2023-03-16-adding-new-llvm-jitlink-target-object-backend/
Congrats again and thank you for the great post!

**Pull request merged**
